src/sink_extra.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2021 Intel Corporation
+ * SPDX-License-Identifier: MIT
+ */
+#include "private.h"
+#include "sink.h"
+
+struct kasan_report_extra_data {
+    emul_insn_t emul_insn;
+    addr_t internal_addr;
+    addr_t internal_offset;
+    addr_t linux_init_task;
+    addr_t linux_kaslr;
+};
+
+/*
+ * Some kernel codes would use kasan_disable_current & kasan_enable_current
+ * to disable & enable the kasan.
+ * Respecting it could prevent the false alarm in the fuzzing report.
+ */
+bool kasan_report_init(vmi_instance_t vmi, struct sink *s) {
+    addr_t vaddr, offset;
+
+    if ( s->extra->data )
+    {
+        printf("Initialized the kasan_report twice\n");
+        return false;
+    }
+
+    struct kasan_report_extra_data *data = malloc(sizeof(struct kasan_report_extra_data));
+    s->extra->data = data;
+
+    data->emul_insn.dont_free = 1;
+    if ( VMI_FAILURE == vmi_read_pa(vmi, s->paddr, 15, &data->emul_insn.data, NULL) )
+    {
+        printf("Failed to read %s PA 0x%lx\n", s->function, s->paddr);
+        return false;
+    }
+
+    if ( VMI_FAILURE == vmi_translate_ksym2v(vmi, "current_task", &vaddr) )
+    {
+        printf("Failed to get current_task\n");
+        return false;
+    }
+    if ( VMI_FAILURE == vmi_get_kernel_struct_offset(vmi, "task_struct", "kasan_depth", &offset) )
+    {
+        printf("Failed to get kasan_depth\n");
+        return false;
+    }
+    if ( VMI_FAILURE == vmi_get_offset(vmi, "linux_kaslr", &data->linux_kaslr) )
+    {
+        printf("Failed to get linux_kaslr\n");
+        return false;
+    }
+    if ( VMI_FAILURE == vmi_get_offset(vmi, "linux_init_task", &data->linux_init_task) )
+    {
+        printf("Failed to get linux_init_task\n");
+        return false;
+    }
+    data->internal_addr = vaddr - data->linux_kaslr;
+    data->internal_offset = offset;
+
+    return true;
+}
+
+sink_cb_response_t kasan_report_cb(vmi_instance_t vmi, vmi_event_t *event, event_response_t *rsp, struct sink *s) {
+    struct kasan_report_extra_data *data = s->extra->data;
+    addr_t current_task;
+    unsigned int kasan_depth;
+    reg_t base;
+
+    if ( !data ) {
+        printf("kasan_report isn't initialized\n");
+        return REPORT_CRASH;
+    }
+
+    if ( VMI_FAILURE == vmi_get_vcpureg(vmi, &base, GS_BASE, 0) )
+    {
+        printf("Failed to call vmi_get_vcpureg 1\n");
+        return REPORT_CRASH;
+    }
+
+    if ( vmi_get_ostype(vmi) == VMI_OS_UNKNOWN)
```

Comment:
Under what scenario would you be getting a kasan callback with OS being unknown?

Reply:
Because we didn't pass the json to `setup_vmi` in `make_fuzz_ready`, the OS type of fuzzdom would be unknown.
Passing the json to `setup_vim` in `make_fuzz_ready` will sacrifice the refork performance, but I think it's a good trade off.